### PR TITLE
Add KVM virtual host support

### PR DESCRIPTION
**Pull request by cbosdo**
Add KVM virtual host module for libvirt. Note that since this is assuming nested KVM works, we won't add the feature to the public cloud providers.

The KVM-based cucumber tests can run on this setup, but not the Xen ones since it can't go nested.

### CODE REVIEW

File: salt/virthost/default-net.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    <name>default</name>
+    <forward mode='nat'>
+      <nat>
+        <port start='1024' end='65535'/>
+      </nat>
+    </forward>
+    <bridge name='virbr0' stp='on' delay='0'/>
+    <ip address='192.168.42.1' netmask='255.255.255.0'>
+      <dhcp>
+        <range start='192.168.42.2' end='192.168.42.254'/>
```
Comment (Bischoff) [inline salt/virthost/default-net.xml]: What if we run two test suites at same time?

What if `192.168.42.*` is already used?

To see how i manage this case, have a look at the handling of `additional_network` keyword.

File: salt/virthost/default-net.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+  <network>
+    <name>default</name>
```
Comment (Bischoff) [inline salt/virthost/default-net.xml]: Hmmm, you will change the user's default libvirt network, right ?

If so, that might be annoying, as he/she might have another use for it.

Wouldn't it make more sense to use a dedicated network?

File: salt/virthost/default-pool.xml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<pool type='dir'>
```
Comment (Bischoff) [inline salt/virthost/default-pool.xml]: Uh, we already have other pools in production (usually called `ssd` IIRC)

If you want a fixed dir, don't use a pool, use filenames?

File: salt/virthost/default-pool.xml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  <name>default</name>
+  <capacity unit='bytes'>0</capacity>
+  <allocation unit='bytes'>0</allocation>
+  <available unit='bytes'>0</available>
+  <source>
+  </source>
+  <target>
+    <path>/var/lib/libvirt/images</path>
```
Comment (Bischoff) [inline salt/virthost/default-pool.xml]: What if we lack space there?

File: salt/controller/bashrc [hunk 5]
```diff
@@ -9,6 +9,10 @@ export MINION="{{ grains.get('minion') }}"
 {% if grains.get('pxeboot_mac') | default(false, true) %}export PXEBOOTMAC="{{ grains.get('pxeboot_mac') }}" {% else %}# no JeOS minion defined {% endif %}
 {% if grains.get('additional_network') | default(false, true) %}export PRIVATENET="{{ grains.get('additional_network') }}" {% else %}# no private network defined {% endif %}
 {% if grains.get('mirror') | default(false, true) %}export MIRROR="yes" {% else %}# no mirror used {% endif %}
+{% if grains.get('kvm_host') | default(false, true) %}
+export VIRTHOST_KVM_URL="{{ grains.get('kvm_host') }}"
+export VIRTHOST_KVM_PASSWORD="linux"
```
Comment (Bischoff) [inline salt/controller/bashrc]: OK, I suppose you need that one because you might be using bare metal virt hosts ?

File: salt/virthost/default-net.xml (hunk 1 above)
Comment (cbosdo) [reply salt/virthost/default-net.xml]: That network is within the virtual host VM... as long as the test suite spawns VMs with different names they shouldn't overlap since those are NAT networks.

File: salt/virthost/default-net.xml (hunk 2 above)
Comment (cbosdo) [reply salt/virthost/default-net.xml]: again, that's the virtual host VM... but yeah I can use a specific name for it.

File: salt/virthost/default-pool.xml (hunk 4 above)
Comment (cbosdo) [reply salt/virthost/default-pool.xml]: Again, this is on the virtual host.... it's a sumaformed VM

File: salt/virthost/default-pool.xml (hunk 3 above)
Comment (cbosdo) [reply salt/virthost/default-pool.xml]: It just works with pools and volumes, and this is in the virtual host sumaformed VM

File: salt/controller/bashrc (hunk 5 above)
Comment (cbosdo) [reply salt/controller/bashrc]: yes, I also run the tests on a sumaform setup without nested virtualization. In such a case I'm setting that variable (and it's Xen friend) manually.

File: salt/virthost/default-pool.xml (hunk 3 above)
Comment (cbosdo) [reply salt/virthost/default-pool.xml]: And yes the test suite is based on pools and volumes.

File: README_ADVANCED.md [hunk 6]
```diff
@@ -730,6 +730,29 @@ module "sumaheadpg" {
 }
 ```
 
+## Virtual hosts
+
+Virtualization host are salt minion hosts that server virtual machines, using KVM hypervisor.
```
Comment (moio) [inline README_ADVANCED.md]: "Virtualization hosts are Salt minions that are also capable to run virtual machines using the KVM hypervisor"

File: README_ADVANCED.md [hunk 7]
```diff
@@ -730,6 +730,29 @@ module "sumaheadpg" {
...
+Note that Xen does not support nested virtualization.
```
Comment (moio) [inline README_ADVANCED.md]: "As this mechanism relies on nested virtualization, Xen is not supported."

File: README_ADVANCED.md [hunk 8]
```diff
@@ -730,6 +730,29 @@ module "sumaheadpg" {
 }
 ```
 
+## Virtual hosts
```
Comment (moio) [inline README_ADVANCED.md]: As this is libvirt-specific, it should go in https://github.com/moio/sumaform/blob/master/modules/libvirt/README.md rather than here.

File: README_ADVANCED.md [hunk 9]
```diff
@@ -730,6 +730,29 @@ module "sumaheadpg" {
...
+  ...
+  name = "min-kvm"
+  image = "sles15sp1"
+  ...
+  vcpu = 3
+  memory = 2048
+}
+```
+
+The created virtual host will get it's host CPU model. This means that in order for virtual hosts to host virtual machines, nested virtualization has to be enabled on the physical machine.
```
Comment (moio) [inline README_ADVANCED.md]: It would be nice to add a pointer on instructions about how to do that, at least in typical cases.

File: main.tf.libvirt-testsuite.example [hunk 10]
```diff
@@ -29,6 +29,7 @@ module "ctl" {
   ubuntu_configuration = "${module.min-ubuntu.configuration}"          // optional
   minionssh_configuration = "${module.minssh-sles12sp3.configuration}" // optional
   pxeboot_configuration = "${module.pxeboot-sles12sp3.configuration}"  // optional
+  kvmhost_configuration = "${module.min-kvm.configuration}"  // optional
```
Comment (moio) [inline main.tf.libvirt-testsuite.example]: Please align comment

File: modules/libvirt/client/variables.tf [hunk 11]
```diff
@@ -90,3 +90,8 @@ variable "mac" {
   description = "a MAC address in the form AA:BB:CC:11:22:22"
   default = ""
 }
+
+variable "cpu_model" {
```
Comment (moio) [inline modules/libvirt/client/variables.tf]: Why is it important to expose this variable, as it is not used neither in the README nor in the example files?

File: modules/libvirt/host/main.tf [hunk 12]
```diff
@@ -23,6 +23,11 @@ resource "libvirt_domain" "domain" {
   count = "${var.count}"
   qemu_agent = true
 
+  // copy host CPU model to guest to get the vmx flag if present
+  cpu {
+    mode = "${var.cpu_model != "" ? "${var.cpu_model}" : "custom"}"
```
Comment (moio) [inline modules/libvirt/host/main.tf]: Shouldn't we use "custom" as the default instead of ""?

File: modules/libvirt/mirror/variables.tf [hunk 13]
```diff
@@ -40,3 +40,8 @@ variable "mac" {
   description = "a MAC address in the form AA:BB:CC:11:22:22"
   default = ""
 }
+
+variable "cpu_model" {
+  description = "Define what CPU model the guest is getting (host-model, host-passthrough or the default)."
```
Comment (moio) [inline modules/libvirt/mirror/variables.tf]: Does this make sense for `mirror` and other modules at all? Why would one want a `mirror` to act as a virtualization host?

File: modules/libvirt/virthost/main.tf [hunk 14]
```diff
@@ -0,0 +1,30 @@
+module "virthost" {
+  source = "../host"
```
Comment (moio) [inline modules/libvirt/virthost/main.tf]: In the README this is documented as being a minion, while here it seems to be based on a vanilla host instead. Which one is correct?

File: salt/virthost/default-net.xml [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+    <forward mode='nat'>
+      <nat>
+        <port start='1024' end='65535'/>
+      </nat>
+    </forward>
+    <bridge name='virbr0' stp='on' delay='0'/>
+    <ip address='192.168.42.1' netmask='255.255.255.0'>
```
Comment (moio) [inline salt/virthost/default-net.xml]: In general, it would be good if we hinted in the README about the virthost's configuration. Something along the lines of: "the virtualization host is set up with a default network of this type with these addresses configured and that default image for guests" - so that whoever uses it gets an idea upfront.

Same goes for the storage pool and other settings.

File: modules/libvirt/virthost/variables.tf [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+variable "hvm_disk_image" {
+  description = "URL to the disk image to use for KVM guests"
+  default = "https://download.opensuse.org/repositories/systemsmanagement:/sumaform:/images:/libvirt/images/opensuse423.x86_64.qcow2"
+}
+
+// Provider-specific variables
+
+variable "image" {
+  description = "One of: sles11sp4, sles12, sles12sp1, sles12sp2, sles12sp3, sles15, centos7"
```
Comment (moio) [inline modules/libvirt/virthost/variables.tf]: Was this really tested on all these OSs, especially CentOS?

File: salt/virthost/init.sls [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+default-pool_undefined:
+  cmd.run:
+    - name: 'virsh pool-undefine default'
+    - onlyif: 'virsh pool-dumpxml default'
+    - require:
+      - cmd: default-pool_destroyed
+
+default-pool_defined:
+  cmd.run:
+    - name: 'virsh pool-define /root/default-pool.xml'
```
Comment (moio) [inline salt/virthost/init.sls]: I wonder if these four pool-related states will work when subsequently running `highstate.sh`. What is the expected behavior, to wipe the pool clean?

File: salt/virthost/init.sls [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    - require:
+      - cmd: default-net_defined
+
+default_virt_net_start:
+  cmd.run:
+    - name: 'virsh net-start default'
+    - require:
+      - cmd: default-net_defined
+
+fake_dmidecode_file:
```
Comment (moio) [inline salt/virthost/init.sls]: Why is this needed? Is it a hack?

File: README_ADVANCED.md (hunk 8 above)
Comment (cbosdo) [reply README_ADVANCED.md]: OK

File: README_ADVANCED.md (hunk 9 above)
Comment (cbosdo) [reply README_ADVANCED.md]: I can document this pretty easily

File: modules/libvirt/virthost/main.tf (hunk 14 above)
Comment (cbosdo) [reply modules/libvirt/virthost/main.tf]: it's a host... but in the test suite it's used as a minion. In order to reuse the minion module I'ld need to have it expose more of the host variables.

File: salt/virthost/default-net.xml (hunk 15 above)
Comment (cbosdo) [reply salt/virthost/default-net.xml]: OK, I'll document that

File: modules/libvirt/virthost/variables.tf (hunk 16 above)
Comment (cbosdo) [reply modules/libvirt/virthost/variables.tf]: I haven't tested on older than sles12sp3 and not on centos, this it just a copy / paste of another module. I can adjust that.

File: salt/virthost/init.sls (hunk 17 above)
Comment (cbosdo) [reply salt/virthost/init.sls]: I could may be use a unless on those: that was mostly to avoid breaking when reapply. The goal is not to wipe: just to ensure the definition is here.

File: salt/virthost/init.sls (hunk 18 above)
Comment (cbosdo) [reply salt/virthost/init.sls]: This is a hack for SUSE Manager / Uyuni, since this one can't stand a VM that is a virtual host. I thus had to completely fake it into a physical machine for it to be properly registered in SUSE Manager.

**Comment by cbosdo** [comment]
@moio About running it in openstack, I'ld rather disable the kvm tests there: that would imply 2 levels of nested virtualization and I'm really unsure how that performs.

File: modules/libvirt/virthost/main.tf (hunk 14 above)
Comment (moio) [reply modules/libvirt/virthost/main.tf]: Please do whatever makes most sense from a general sumaform user's point of view. Of course we want to have clean, maintainable testsuite code as much as possible, but when we add general sumaform features we also want to make them as generally useful as possible.

Maybe in this case really using host gives the maximum flexibility. Not considering the testsuite use case, what would you imagine being more useful?

File: modules/libvirt/virthost/variables.tf (hunk 16 above)
Comment (moio) [reply modules/libvirt/virthost/variables.tf]: I would suggest to restrict this to OSs known to work

**Comment by moio** [comment]
@cbosdo if it's easy and should work (even if slow) I would give that a try anyway - there is a long-term evaluation effort to move more workloads on ECP, and this could be a blocking factor.

File: salt/virthost/init.sls (hunk 17 above)
Comment (cbosdo) [reply salt/virthost/init.sls]: oh and BTW, those four pool-relate states do work after reapplying: they have been written for that ;)

**Comment by cbosdo** [comment]
@moio That wouldn't be a blocking factor as long as we can have a different main.tf for ECP than libvirt test suites. If that module isn't added, then the virtualization tests are all skipped just like now.

**Comment by cbosdo** [comment]
and I had reports of unreliability of nested virt on openstack: clearly not a priority to spend time on it
